File: src/main/scala/my/spark/examples/AverageMovieRatings.scala
package my.spark.examples

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._

object AverageMovieRatings {
  def main(args: Array[String]) {

    val sc = new SparkContext()
  
    val movies =  sc.textFile("hdfs://localhost/user/training/movie/part*")
      .map(_.split("\t"))
      .map(tokens => (tokens(0),tokens(1)))

    val ratings = sc.textFile("hdfs://localhost/user/training/movierating/part*")
      .map(_.split("\t"))
      .map(tokens => (tokens(1),tokens(2).toInt))
      .groupByKey()
      .mapValues(ratings => ratings.sum/ratings.size.toFloat)

    val namesRatings = movies.join(ratings)
      .map{case(id,(title,rating)) => (id.toInt,(title,rating))}
      .sortByKey(true)

    namesRatings.take(10).foreach(println)
  
    sc.stop()
  }
}

File: src/main/python/AverageMovieRatings.py
import sys
from pyspark import SparkContext

if __name__ == "__main__":

  sc = SparkContext()
  
  movies =  sc.textFile("hdfs://localhost/user/training/movie/part*")\
    .map(lambda line: line.split("\t"))\
    .map(lambda tokens: (tokens[0],tokens[1]))

  ratings = sc.textFile("hdfs://localhost/user/training/movierating/part*")\
    .map(lambda line: line.split("\t"))\
    .map(lambda tokens: (tokens[1],int(tokens[2])))\
    .groupByKey()\
    .mapValues(lambda ratings: sum(ratings)/float(len(ratings)))

  namesRatings = movies.join(ratings)\
    .map(lambda (id,(title,rating)):(int(id),(title,rating)))\
    .sortByKey(True)

  for (id,(title,rating)) in namesRatings.take(10): print("(" + str(id) + "(" + title + "," + str(rating) + "))")
  
  sc.stop()